src/services/crawl-dids/did_processor.service.ts

```diff
@@ -0,0 +1,236 @@
+    type: string;
+    did?: string;
+    attributes?: DidAttribute[];
+    [key: string]: any;
+}
+
+interface DidMessageTypes {
+    event_type: string;
+    did?: string;
+    created?: string;
+    modified?: string;
+    deleted_at?: string | null;
+    height?: number;
+    is_deleted?: boolean;
+    exp?: string;
+    deposit?: string;
+    years?: string;
+    id?: string;
+    changes?: any;
+    [key: string]: any;
+}
+
+@Service({
+    name: SERVICE.V1.DidMessageProcessorService.key,
+    version: 1,
+})
+export default class DidMessageProcessorService extends BullableService {
+    constructor(broker: ServiceBroker) {
+        super(broker);
+    }
+
+
+    private computeChanges(
+        oldRec: Record<string, any>,
+        newRec: Record<string, any>
+    ): Record<string, { old: any; new: any }> {
+        const diff: Record<string, { old: any; new: any }> = {};
+        Object.keys(newRec).forEach((key) => {
+            if (newRec[key] !== oldRec[key]) {
+                diff[key] = { old: oldRec[key], new: newRec[key] };
+            }
+        });
+        return diff;
+    }
+
+
+
+    private async saveHistory(didEvent: DidMessageTypes, changes?: any) {
+        const { modified, ...cleanedEvent } = didEvent;
+        await this.broker.call(`${SERVICE.V1.DidHistoryService.path}.save`, {
+            ...cleanedEvent,
+            changes: changes ? JSON.stringify(changes) : null,
+        });
+    }
+
+
+    @Action({ name: "handleDidMessages" })
+    async handleDidMessages(ctx: { params: { listDidTx: DidEvent[] } }) {
+        const { listDidTx } = ctx.params;
+
+        for (const event of listDidTx) {
+            let processedEvent: DidMessageTypes | null = null;
+            const calculateDeposit = await calculateDidDeposit();
+            // ---------------- ADD ----------------
+            if ([DidEventTypes.AddDid, DidEventTypes.AddDidLegacy].includes(event.type as DidEventTypes)) {
+                processedEvent = {
+                    event_type: event.type,
+                    did: event.did,
+                    controller: event.controller,
+                    height: event.height ?? 0,
+                    years: event.years ? String(event.years) : undefined,
+                    deposit: String(calculateDeposit) ?? "0",
+                    created: formatTimestamp(event?.timestamp),
+                    modified: formatTimestamp(event?.timestamp),
+                    exp:
+                        event.timestamp && event.years
+                            ? addYearsToDate(event.timestamp, event.years)
+                            : undefined,
+                    is_deleted: false,
+                    deleted_at: null,
+                };
+
+
+                event.attributes?.forEach((attr: DidAttribute) => {
+                    if (!["timestamp", "height"].includes(attr.key)) {
+                        processedEvent![attr.key] = attr.value;
+                    }
+                });
```

Comment:
Does this mean that we will store any arbitrary key-value that a transaction could contain? Wouldn't be better to be strict on the data model of our tables?